[BOLT] Refactor MCInstReference and move it to Core (NFC)

atrosinenko opened the pull request:
Refactor MCInstReference class and move it from PAuth gadget scanner to Core.

MCInstReference is a class representing a constant reference to an instruction inside a parent entity - either inside a basic block (which has a reference to its parent function) or directly inside a function (when CFG information is not available).

atrosinenko commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/138655?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#138655** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/138655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/138655?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#137975** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/137975?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#137224** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/137224?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#136183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136183?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#136151** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136151?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#135663** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135663?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#136147** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136147?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#135662** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135662?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#135661** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135661?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134146** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134146?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133461** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133461?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#135073** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/135073?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

atrosinenko commented:
This is similar to `MCPlusBuilder::InstructionIterator` class, but not tied to some specific iteration semantics. In a subsequent PR I plan adding an `std::optional<MCInstReference> getSinglePredecessor()` function with an extra feature of handling the case of the first instruction in a basic block with exactly one predecessor (this is important to analyze pauth-specific `BR_JumpTable` lowering), but other flavors of `operator--` could be possible.

maksfb commented on bolt/include/bolt/Core/MCInstUtils.h:
```diff
@@ -0,0 +1,168 @@
+    const BinaryBasicBlock *BB;
+    BinaryBasicBlock::const_iterator It;
+
+    bool operator<(const RefInBB &Other) const {
```
What are expected uses for this comparison? I'm concerned about non-deterministic order of `BinaryBasicBlock *`.

maksfb commented on bolt/include/bolt/Core/MCInstUtils.h:
```diff
@@ -0,0 +1,168 @@
+    const BinaryFunction *BF;
+    nocfg_const_iterator It;
+
+    bool operator<(const RefInBF &Other) const {
```
Similar concern regarding the `BinaryFunction *` order.

maksfb commented on bolt/lib/Passes/PAuthGadgetScanner.cpp:
```diff
@@ -1682,48 +1648,66 @@ void Analysis::runOnFunction(BinaryFunction &BF,
+    auto It = static_cast<BinaryBasicBlock::const_iterator>(&Inst.getMCInst());
+    unsigned IndexInBB = std::distance(BB->begin(), It);
+
+    // FIXME: this assumes all instructions are 4 bytes in size. This is true
```
We have `BinaryContext::computeCodeSize()`.

maksfb commented on bolt/lib/Core/MCInstUtils.cpp:
```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+
+#include "bolt/Core/MCInstUtils.h"
+
```
nit: empty line.

maksfb reviewed (approved):
As an NFC this change looks good to me. I've left a comments for a follow-up.

atrosinenko committed:
[BOLT] Factor out MCInstReference from gadget scanner (NFC)

Move MCInstReference representing a constant reference to an instruction
inside a parent entity - either inside a basic block (which has a
reference to its parent function) or directly to the function (when CFG
information is not available).

atrosinenko committed:
Resolve review comments

atrosinenko reviewed (commented):
@maksfb Thank you for the comments, kept switching to `BinaryContext::computeCodeSize()` for a follow-up, addressed other comments in e5a9182baef7785429b17a41748d8bd89b86ab0b.

Updated the title and the description of this PR, as the original ones were harder to understand.

atrosinenko committed:
Last-minute fixes

atrosinenko commented:
Added a few last-minute fixes in 5e906b0c6a90f1eb9b596bb7c8d0fb1a1ca27320: dropped unused `#include`, added a few assertions (to make sure various getters are not called on an empty reference), but most importantly replaced an implicit `operator bool()` with explicit `MCInstReference::empty()` function, as it turned out implicit conversion to `bool` unintentionally made `MCInstReference`s comparable.

atrosinenko commented:
@nico Thank you for pointing this out, I have to come up with a better way to handle iterators / pointers to `MCInst *` without **implicitly** relying on BinaryBasicBlock storing its instructions in a vector, so reverting this PR for now.
